api/src/test/java/org/apache/iceberg/types/TestNestedFieldDefaultValues.java

```diff
@@ -0,0 +1,77 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.types;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+
+public class TestNestedFieldDefaultValues {
+
+  @Test
+  public void testConstructors() {
+    int id = 1;
+    String fieldName = "fieldName";
+    Type fieldType = Types.IntegerType.get();
+    String doc = "field doc";
+    Integer defaultValue = 100;
+
+    // optional constructors
+    Assert.assertFalse(Types.NestedField.optional(id, fieldName, fieldType).hasDefaultValue());
+    Assert.assertFalse(Types.NestedField.optional(id, fieldName, fieldType, doc).hasDefaultValue());
+    Types.NestedField nestedFieldWithDefault = Types.NestedField.optional(id, fieldName, fieldType, defaultValue, doc);
+    Assert.assertTrue(nestedFieldWithDefault.hasDefaultValue());
+    Assert.assertEquals(defaultValue, nestedFieldWithDefault.getDefaultValue());
+    nestedFieldWithDefault = Types.NestedField.optionalWithDefault(id, fieldName, fieldType, defaultValue);
+    Assert.assertTrue(nestedFieldWithDefault.hasDefaultValue());
+    Assert.assertEquals(defaultValue, nestedFieldWithDefault.getDefaultValue());
+
+    // required constructors
+    Assert.assertFalse(Types.NestedField.required(id, fieldName, fieldType).hasDefaultValue());
+    Assert.assertFalse(Types.NestedField.required(id, fieldName, fieldType, doc).hasDefaultValue());
+    nestedFieldWithDefault = Types.NestedField.required(id, fieldName, fieldType, defaultValue, doc);
+    Assert.assertTrue(nestedFieldWithDefault.hasDefaultValue());
+    Assert.assertEquals(defaultValue, nestedFieldWithDefault.getDefaultValue());
+    nestedFieldWithDefault = Types.NestedField.requiredWithDefault(id, fieldName, fieldType, defaultValue);
+    Assert.assertTrue(nestedFieldWithDefault.hasDefaultValue());
+    Assert.assertEquals(defaultValue, nestedFieldWithDefault.getDefaultValue());
+
+    // of constructors
+    Assert.assertFalse(Types.NestedField.of(id, true, fieldName, fieldType).hasDefaultValue());
+    Assert.assertFalse(Types.NestedField.of(id, true, fieldName, fieldType, doc).hasDefaultValue());
+    nestedFieldWithDefault = Types.NestedField.of(id, true, fieldName, fieldType, defaultValue, doc);
+    Assert.assertTrue(nestedFieldWithDefault.hasDefaultValue());
+    Assert.assertEquals(defaultValue, nestedFieldWithDefault.getDefaultValue());
+
+    // illegal case (required with null defaultValue
+    try {
```

Comment:
Can we pull these illegal cases into their separate test methods? We can use JUnit's [`ExpectedException`](https://junit.org/junit4/javadoc/4.12/org/junit/rules/ExpectedException.html) to ensure that the exception class and the message are what we expect.

Reply:
I would agree execept I think we would want to use our custom AssertHelpers.
https://github.com/apache/iceberg/blob/master/api/src/test/java/org/apache/iceberg/AssertHelpers.java#L38

Rather than the Junit code